File: cryptography/dotnet/Helverify.Cryptography/ZeroKnowledge/ProofOfContainingOne.cs
using Helverify.Cryptography.Helper;
using Org.BouncyCastle.Crypto.Digests;
using Org.BouncyCastle.Crypto.Prng;
using Org.BouncyCastle.Math;
using Org.BouncyCastle.Security;

namespace Helverify.Cryptography.ZeroKnowledge
{
    /// <summary>
    /// Bernhard2016
    /// KillerProvotum2020
    /// Chaum1993
    ///
    /// Variation of DCP proof adapted towards only proving that the ciphertext contains 1.
    /// => useful for row and column product == 1 proof
    /// </summary>
    public class ProofOfContainingOne
    {
        private static readonly SecureRandom SecureRandom = new(new DigestRandomGenerator(new Sha256Digest()));

        public BigInteger U { get; }
        public BigInteger V { get; }
        public BigInteger S { get; }

        /// <summary>
        /// Private constructor, use <see cref="Create"/> instead.
        /// </summary>
        /// <param name="u">Proof parameter u</param>
        /// <param name="v">Proof parameter v</param>
        /// <param name="s">Proof parameter s</param>
        private ProofOfContainingOne(BigInteger u, BigInteger v, BigInteger s)
        {
            U = u;
            V = v;
            S = s;
        }

        /// <summary>
        /// Creates a proof that a ciphertext contains the value one.
        /// </summary>
        /// <param name="a">First component of an ElGamal ciphertext</param>
        /// <param name="b">Second component of an ElGamal ciphertext</param>
        /// <param name="h">Public key of an ElGamal cryptosystem</param>
        /// <param name="x">Private key of an ElGamal cryptosystem</param>
        /// <param name="p">Public prime p of an ElGamal cryptosystem</param>
        /// <param name="g">Generator g of an ElGamal cryptosystem</param>
        /// <returns>Proof that a ciphertext contains the value one.</returns>
        public static ProofOfContainingOne Create(BigInteger a, BigInteger b, BigInteger h, BigInteger x, BigInteger p, BigInteger g)
        {
            BigInteger q = p.Subtract(BigInteger.One).Multiply(BigInteger.Two.ModInverse(p)).Mod(p);

            BigInteger r = new BigInteger(x.BitLength, SecureRandom).Mod(q);

            BigInteger u = g.ModPow(r, p).Mod(p);
            BigInteger v = h.ModPow(r, p).Mod(p);

            BigInteger c = HashHelper.GetHash(q, h, a, b, u, v).Mod(q);

            BigInteger s = r.Add(c.Multiply(x).Mod(q)).Mod(q);

            return new ProofOfContainingOne(u, v, s);
        }

        /// <summary>
        /// Allows to verify that a ciphertext contains the value one.
        /// </summary>
        /// <param name="a">First component of an ElGamal ciphertext</param>
        /// <param name="b">Second component of an ElGamal ciphertext</param>
        /// <param name="h">Public key of an ElGamal cryptosystem</param>
        /// <param name="p">Public prime p of an ElGamal cryptosystem</param>
        /// <param name="g">Generator g of an ElGamal cryptosystem</param>
        /// <returns>True if the ciphertext represents one, false otherwise.</returns>
        public bool Verify(BigInteger a, BigInteger b, BigInteger h, BigInteger p, BigInteger g)
        {
            BigInteger q = p.Subtract(BigInteger.One).Multiply(BigInteger.Two.ModInverse(p)).Mod(p);

            BigInteger c = HashHelper.GetHash(q, h, a, b, U, V).Mod(q);

            bool firstCondition = g.ModPow(S, p).Mod(p).Equals(U.Multiply(a.ModPow(c, p)).Mod(p));

            BigInteger gInverse = g.ModInverse(p).Mod(p);

            bool secondCondition = h.ModPow(S, p).Mod(p).Equals(V.Multiply(b.Multiply(gInverse).Mod(p).ModPow(c, p).Mod(p)).Mod(p));

            return firstCondition && secondCondition;
        }
    }
}

File: voter/frontend/src/cryptography/proofOfContainingOne.ts
import bigInt from "big-integer";
import {HashHelper} from "../helper/hashHelper";
import {BigNumberHelper} from "../helper/bigNumberHelper";

/**
 * Proves that a ciphertext (or a sum of ciphertexts) contains the value 1
 */
export class ProofOfContainingOne{

    /**
     * U component
     */
    u: bigInt.BigInteger;

    /**
     * V component
     */
    v: bigInt.BigInteger;

    /**
     * S component
     */
    s: bigInt.BigInteger;

    /**
     * Constructor
     * @param uHex U component as a hexadecimal string
     * @param vHex V component as a hexadecimal string
     * @param sHex S component as a hexadecimal string
     */
    constructor(uHex: string, vHex: string, sHex: string) {
        this.u = BigNumberHelper.fromHexString(uHex);
        this.v = BigNumberHelper.fromHexString(vHex);
        this.s = BigNumberHelper.fromHexString(sHex);
    }

    /**
     * Verifies that the proof is valid, i.e., that the ciphertext indeed contains the value 1
     * @param a C component of the ElGamal ciphertext
     * @param b D component of the ElGamal ciphertext
     * @param h Election public key
     * @param p ElGamal parameter p
     * @param g ElGamal parameter g
     */
    verify(a: bigInt.BigInteger, b: bigInt.BigInteger, h: bigInt.BigInteger, p: bigInt.BigInteger, g: bigInt.BigInteger): boolean{
        const q: bigInt.BigInteger = p.subtract(bigInt.one).multiply(bigInt(2).modInv(p)).mod(p);

        const c: bigInt.BigInteger = HashHelper.getHash(q, [h, a, b, this.u, this.v]);

        const gInverse: bigInt.BigInteger = g.modInv(p).mod(p);

        const condition1 : boolean = g.modPow(this.s, p).mod(p).equals(this.u.multiply(a.modPow(c, p)).mod(p));

        const condition2 : boolean = h.modPow(this.s, p).equals(this.v.multiply(b.multiply(gInverse).mod(p).modPow(c, p).mod(p)).mod(p));

        return condition1 && condition2;
    }
}